kotlin-native/litmus/src/commonMain/kotlin/BasicLitmusTest.kt

```diff
@@ -0,0 +1,104 @@
+import kotlin.native.concurrent.AtomicReference
+import kotlin.reflect.KClass
+import kotlin.system.getTimeMillis
+import kotlin.time.Duration
+
+abstract class BasicLitmusTest(val name: String) {
+    abstract suspend fun actor1(): Any?
+    abstract suspend fun actor2(): Any?
+    open suspend fun actor3(): Any? = Placeholder
+    open suspend fun actor4(): Any? = Placeholder
+    open suspend fun actor5(): Any? = Placeholder
+    open suspend fun actor6(): Any? = Placeholder
+    open suspend fun actor7(): Any? = Placeholder
+    open suspend fun actor8(): Any? = Placeholder
+    open fun arbiter() {}
+
+    private val outcomeRef = AtomicReference<Any?>(null)
+    var outcome: Any?
+        protected set(v) {
+            if (!outcomeRef.compareAndSet(null, v))
+                throw IllegalStateException("cannot set outcome more than once")
+        }
+        get() = outcomeRef.value
+
+    // note: modifies receiver object
+    suspend fun overriddenActors(): List<suspend (BasicLitmusTest) -> Any?> {
+        val actors = mutableListOf(BasicLitmusTest::actor1, BasicLitmusTest::actor2)
+        if(actor3() != Placeholder) actors.add(BasicLitmusTest::actor3)
+        if(actor4() != Placeholder) actors.add(BasicLitmusTest::actor4)
+        if(actor5() != Placeholder) actors.add(BasicLitmusTest::actor5)
+        if(actor6() != Placeholder) actors.add(BasicLitmusTest::actor6)
+        if(actor7() != Placeholder) actors.add(BasicLitmusTest::actor7)
+        if(actor8() != Placeholder) actors.add(BasicLitmusTest::actor8)
+        return actors
+    }
+
+    companion object {
+        var memShuffler: MemShuffler? = null
+
+        private object Placeholder
+    }
+}
+
+data class OutcomeSetupScope(
+        var accepted: Set<Any?> = emptySet(),
+        var interesting: Set<Any?> = emptySet(),
+        var forbidden: Set<Any?> = emptySet(),
+        var forbidOther: Boolean = false,
+) {
+    fun getType(outcome: Any?) = when {
+        outcome in accepted -> OutcomeType.ACCEPTED
+        outcome in interesting -> OutcomeType.INTERESTING
+        outcome in forbidden || forbidOther -> OutcomeType.FORBIDDEN
+        else -> null
+    }
+}
+
+private val testOutcomesSetup = mutableMapOf<KClass<*>, OutcomeSetupScope>()
+
+fun BasicLitmusTest.setupOutcomes(block: OutcomeSetupScope.() -> Unit) {
+    testOutcomesSetup[this::class] = OutcomeSetupScope().apply(block)
+}
+
+fun BasicLitmusTest.getOutcomeSetup(): OutcomeSetupScope? = testOutcomesSetup[this::class]
+
+typealias AffinityMap = List<Set<Int>>
```

Comment:
Perhaps, making it a class instead of just a type alias would make more sense, because logically affinity map is more than just list of sets of thread identifiers.